symengine/symengine_rcp.h

```diff
@@ -198,6 +198,73 @@ using Teuchos::print_stack_on_segfault;
 #endif
 
 
+template <class T>
+class EnableRCPFromThis {
+// Public interface
+public:
+    //! Get RCP<T> pointer to self (it will cast the pointer to T)
+    inline RCP<T> rcp_from_this() {
+#if defined(WITH_SYMENGINE_RCP)
+        return rcp(static_cast<T*>(this));
+#else
+        return rcp_static_cast<T>(weak_self_ptr_.create_strong());
+#endif
+    }
+
+    //! Get RCP<const T> pointer to self (it will cast the pointer to const T)
+    inline RCP<const T> rcp_from_this() const {
```

Comment:
I used the same approach as in the standard library:

http://en.cppreference.com/w/cpp/memory/enable_shared_from_this/shared_from_this

Which also has the two methods, one returning `const T`, the other just `T`. One thing that I don't understand is that it seems the method itself never modifies the original object, so it seems both of the methods can be const. But for this PR I am just trying to follow the standard.

In the long run, @bartlettroscoe and I were thinking of trying to get the C++ standard extended, so that the `unique_ptr`, `shared_ptr` and `weak_ptr` can also return a `Ptr` class (it would probably be just called `ptr` or `view_ptr`) using a new `.ptr()` method, as opposed to a raw pointer as it is now using the `.get()` method. For self reference, one has to use `std::enable_shared_from_this`, and so our `EnableRCPFromThis` is a safe counterpart, and so I am trying to follow the same API. That way anybody using just pure C++ could use these classes. For now we have to implement these ourselves.